From d11fc30e0a7c51a2ef048b8119260feb0823defc Mon Sep 17 00:00:00 2001
From: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>
Date: Mon, 25 Aug 2025 18:32:16 +0530
Subject: [PATCH] Adds list command to cephfs mirroring

This patch adds a CLI command to support listing active paths enabled
for mirroring in a volume.

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>
---
 /share/ceph/mgr/mirroring/fs/snapshot_mirror.py | 14 ++++++++++++++
 /share/ceph/mgr/mirroring/module.py             |  6 ++++++
 2 files changed, 20 insertions(+)

diff --git a/share/ceph/mgr/mirroring/fs/snapshot_mirror.py b/share/ceph/mgr/mirroring/fs/snapshot_mirror.py
index 2bfb6482674..c348ce82de1 100644
--- a/share/ceph/mgr/mirroring/fs/snapshot_mirror.py
+++ b/share/ceph/mgr/mirroring/fs/snapshot_mirror.py
@@ -722,6 +722,20 @@ class FSSnapshotMirror:
         except Exception as e:
             return e.args[0], '', 'failed to remove directory'
 
+    def list_dirs(self, filesystem):
+        try:
+            with self.lock:
+                if not self.filesystem_exist(filesystem):
+                    raise MirrorException(-errno.ENOENT, f'filesystem {filesystem} does not exist')
+                fspolicy = self.pool_policy.get(filesystem, None)
+                if not fspolicy:
+                    raise MirrorException(-errno.EINVAL, f'filesystem {filesystem} is not mirrored')
+                return 0, json.dumps(list(fspolicy.policy.dir_states.keys()), indent=4, sort_keys=True), ''
+        except MirrorException as me:
+            return me.args[0], '', me.args[1]
+        except Exception as e:
+            return e.args[0], '', 'failed to list directories'
+
     def status(self,filesystem, dir_path):
         try:
             with self.lock:
diff --git a/share/ceph/mgr/mirroring/module.py b/share/ceph/mgr/mirroring/module.py
index 4b4354ab2b9..5ba7dde5587 100644
--- a/share/ceph/mgr/mirroring/module.py
+++ b/share/ceph/mgr/mirroring/module.py
@@ -84,6 +84,12 @@ class Module(MgrModule):
         """Remove a snapshot mirrored directory"""
         return self.fs_snapshot_mirror.remove_dir(fs_name, path)
 
+    @CLIReadCommand('fs snapshot mirror ls')
+    def snapshot_mirror_ls(self,
+                           fs_name: str):
+        """List the snapshot mirrored directories"""
+        return self.fs_snapshot_mirror.list_dirs(fs_name)
+
     @CLIReadCommand('fs snapshot mirror dirmap')
     def snapshot_mirror_dirmap(self,
                                fs_name: str,
-- 
2.43.0

